feat: Add data-lineage overview (#1383)
seqeralabs/docs: platform-cloud/docs/data/data-lineage.md

- **Tasks**: Displays the lineage ID and lineage labels for each `TaskRun` alongside existing task data, so you can trace any task back to its lineage record. All task file inputs and outputs, and upstream and downstream tasks linked by lineage records, are displayed.
- **Inputs**: Lists all input datasets and parameters with file paths, types, and lineage IDs and lineage labels where available.
- **Outputs**: Lists all `FileOutput` records linked to the workflow run: output name, file path, type, lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].

:::tip
All LIDs and lineage labels are clickable links. Click any LID to open the organization-level lineage search pre-filled with that identifier.
:::

### Data Explorer

Output objects from a lineage-enabled run display their LID and any lineage labels when you preview the object in Data Explorer. You can trace any file back to the pipeline run that produced it.

## Lineage labels

Assign lineage labels to output files using the `label` directive in your Nextflow process definitions. Labels appear in lineage records and are searchable across your workspace.

> Both Seqera Platform labels and Nextflow lineage labels propagate to lineage records. Seqera Platform excludes resource labels as they relate to underlying compute resources, not the data itself.

Reviewer: Do Seqera Platform labels propagate to lineage records?
I tried this and I don't think they do propagate to the record (only the Nextflow lineage labels were there as far as I can see)

Author: Will test to make sure

Author: Per @bentsherman , provided the Nextflow version used is `26.04`, then Seqera Platform labels propagate to the lineage WorkflowRun record.

Author: And according to @swampie, the default Nextflow version in `v26.1` will be `26.04`, so it will be supported.

---
title: "Data Lineage"
description: "Using data lineage in Seqera Platform."
date created: "2026-05-04"
last updated: "2026-05-04"
tags: [data lineage, provenance, governance, reproducibility, lineage id, lid, label]
---

:::info
Data lineage in Platform is in public preview. It requires Nextflow v25.04 or later.
:::

:::warning
The feature is experimental and subject to change. See this guide for the latest configuration recommendations and limitations.
:::

Data lineage tracks the full provenance of every pipeline run at both the task and workflow level, including what executed, what data it consumed, and what outputs it produced. Use it to audit results, verify reproducibility, and trace file provenance.

> ## Why data lineage matters

Reviewer 2: ```suggestion
## Overview
```

Production pipelines generate results that teams need to trust, audit, and reproduce. Data lineage provides a precise, immutable record of how each result was produced.

- **Reproducibility**: Every run, task, and output file receives a unique lineage ID (LID), a traversable URI that points to a structured record of what ran. Verify that two runs produced identical results, or identify where they diverged.
- **Auditing and compliance**: For teams in regulated industries such as pharma, clinical genomics, and CROs, lineage provides the audit trail needed for regulatory compliance. Each record captures inputs, outputs, parameters, compute environment, and the user who launched the run.
- **Debugging**: When a cached task unexpectedly re-executes, or a pipeline produces an unexpected result, lineage traces backward from any output to all contributing tasks and parameters. Compare two task runs to isolate what changed.
- **Broader team access**: Exploring Nextflow lineage previously required CLI access and comfort reading raw JSON. Platform now surfaces lineage data in pipeline run detail pages and Data Explorer. Users can inspect provenance directly.

> - **Cross-workflow discoverability**: [Workflow output labels][workflow-labels] make output files discoverable across runs. Query lineage records by label to find all matching outputs workspace-wide, without knowing which specific run produced a file.

Reviewer 3: >  Query lineage records by label to find all matching outputs workspace-wide

Can you? How do you do this?

Author: Fixed.

## How data lineage works

Nextflow creates a structured JSON record for each entity in your pipeline when lineage is enabled:

| Record type | Description |
|---|---|
| **WorkflowRun** | Full pipeline execution: repository, commit ID, parameters, compute environment, session ID, and Platform context (user, workspace, pipeline) |
| **TaskRun** | Individual task execution: script, code checksum, inputs, outputs, container, and dependencies |
| **FileOutput** | Output file: path, checksum, size, timestamp, and links back to the task and workflow that produced it |

> Each record gets a lineage ID (LID), a `lid://` URI that uniquely identifies the entity. Every LID and lineage label renders as a clickable link, letting you navigate to all related entities across your organization.

Reviewer 2: ```suggestion
Each record gets a lineage ID (LID), a `lid://` URI that uniquely identifies the entity. Every LID and lineage label renders as a clickable link, and you can navigate to all related entities across your organization.
```

### Configure workspace settings

> Before collecting lineage data, configure the lineage storage location for your workspace. Go to **Workspace Settings** and open the [**Lineage**][workspace-lineage] settings to set the storage bucket and path where lineage data is stored and indexed. This applies to all pipeline runs in the workspace.

Reviewer 2: ```suggestion
Before collecting lineage data, configure the lineage storage location for your workspace. Go to **Settings > Workspace Settings**. Select **Lineage** to set the storage bucket and path where lineage data is stored and indexed. This applies to all pipeline runs in the workspace. See [Lineage][workspace-lineage] for more information about the settings.
```

> Before collecting lineage data, configure the lineage storage location for your workspace. Go to **Workspace Settings** and open the [**Lineage**][workspace-lineage] settings to set the storage bucket and path where lineage data is stored and indexed. This applies to all pipeline runs in the workspace.

Reviewer 3: This talks about location, but not about enabling lineage at workspace-level (which is what we expect / want most people to do).

Author: Fixed.

:::danger
Changing the lineage storage bucket path after lineage data is generated results in historic data loss. The lineage index is tied to this location. Changing it makes existing records inaccessible. To move the storage location, first copy all existing lineage data to the new bucket and path (for example, `aws s3 cp --recursive s3://old-bucket/path s3://new-bucket/path`), then update the workspace setting.
:::

### Enable per pipeline lineage in Nextflow

> To test lineage within a single pipeline, add the following to your Nextflow config file before running your pipeline:

Reviewer 3: Related to above - I think this should be the exception rather than the rule. We should make this clearer.

Also, you can enable it in the launchpad for a single run / pipeline, within the Platform UI - no need to toggle `enabled` in the Nextflow config like this. No mention of that in these docs.

Author: Fixed.

> ### Enable per pipeline lineage in Nextflow

Reviewer 4: ````suggestion
:::

#### AWS IAM permissions

Lineage requires additional IAM permissions beyond the standard set documented for AWS Batch and AWS Cloud compute environments. If you use an existing AWS Batch queue or AWS Cloud compute environment with custom IAM roles, update both the service role and Seqera Platform integration credentials with the policies below.

:::info
The examples below assume the default lineage bucket naming convention `seqera-lineage-<workspace-id>`. Replace `<workspace-id>` with your workspace ID, or adjust the resource ARNs to match the bucket configured in your lineage workspace.
:::

**Service role**

Add the following policy to the service role used by your AWS Batch queue or Cloud compute environment:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ListObjectsInBucket",
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket"
            ],
            "Resource": "arn:aws:s3:::seqera-lineage-"
        },
        {
            "Sid": "AllObjectActions",
            "Effect": "Allow",
            "Action": "s3:*Object",
            "Resource": "arn:aws:s3:::seqera-lineage-/*"
        },
        {
            "Sid": "AllowObjectTagging",
            "Effect": "Allow",
            "Action": [
                "s3:PutObjectTagging",
                "s3:GetObjectTagging"
            ],
            "Resource": "arn:aws:s3:::seqera-lineage-/*"
        }
    ]
}
```

**Platform integration credentials**

Add the following policy to the AWS credentials used by Seqera Platform. These permissions allow Platform to provision the SQS queue and S3 bucket notifications that feed the lineage index:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "sqs:CreateQueue",
                "sqs:GetQueueAttributes",
                "sqs:SetQueueAttributes",
                "sqs:GetQueueUrl",
                "sqs:ReceiveMessage",
                "sqs:DeleteMessage"
            ],
            "Resource": "arn:aws:sqs:*:*:seqera-lineage-*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "s3:CreateBucket",
                "s3:GetBucketNotificationConfiguration",
                "s3:PutBucketNotificationConfiguration",
                "s3:GetBucketLocation"
            ],
            "Resource": "arn:aws:s3:::seqera-lineage-*"
        }
    ]
}
```

### Enable per pipeline lineage in Nextflow
````
cc @gavinelder

Reviewer 5: @christopher-hakkaart there was an inclusion earlier https://github.com/seqeralabs/docs/pull/1361/changes#diff-5d72ca12da70c5ea15a85c69a02309456c6220831be3631c4628043ffe091239 from @georgi-seqera 

happy to resolve and take your guidance on content.

Reviewer 4: I missed that one, sorry - ignore this. It made me a bit uncomfortable having this here out of context anyway. Will resolve. Thanks!

- **Reproducibility**: Every run, task, and output file receives a unique **lineage ID (LID)** — a traversable URI pointing to a structured record of exactly what ran. You can verify that two runs produced identical results, or identify precisely where they diverged.
- **Auditing and compliance**: For teams in regulated industries (pharma, clinical genomics, CROs), lineage provides the audit trail needed for regulatory compliance. Each record captures inputs, outputs, parameters, compute environment, and the user who launched the run.
- **Debugging**: When a cached task unexpectedly re-executes or a pipeline produces an unexpected result, lineage lets you trace backward from any output to all contributing tasks and parameters. You can compare two task runs to isolate exactly what changed.
- **Broader team access**: Previously, exploring Nextflow lineage required CLI access and comfort reading raw JSON. Lineage data is surfaced in both pipeline run details pages and Data Explorer, so users can inspect provenance directly.

> - **Cross-workflow discoverability**: [Workflow output labels][workflow-labels] make output files discoverable across runs. Rather than knowing which specific run produced a file, query lineage records by label to find all matching outputs workspace-wide.

Reviewer 4: ```suggestion
- **Reproducibility**: Every run, task, and output file receives a unique lineage ID (LID), a traversable URI that points to a structured record of what ran. Verify that two runs produced identical results, or identify where they diverged.
- **Auditing and compliance**: For teams in regulated industries such as pharma, clinical genomics, and CROs, lineage provides the audit trail needed for regulatory compliance. Each record captures inputs, outputs, parameters, compute environment, and the user who launched the run.
- **Debugging**: When a cached task unexpectedly re-executes, or a pipeline produces an unexpected result, lineage traces backward from any output to all contributing tasks and parameters. Compare two task runs to isolate what changed.
- **Broader team access**: Exploring Nextflow lineage previously required CLI access and comfort reading raw JSON. Platform now surfaces lineage data in pipeline run detail pages and Data Explorer. Users can inspect provenance directly.
- **Cross-workflow discoverability**: [Workflow output labels][workflow-labels] make output files discoverable across runs. Query lineage records by label to find all matching outputs workspace-wide, without knowing which specific run produced a file.
```

## How data lineage works

Nextflow creates a structured JSON record for each entity in your pipeline when lineage is enabled:

| Record type | Description |
|---|---|
| **WorkflowRun** | Full pipeline execution: repository, commit ID, parameters, compute environment, session ID, and Platform context (user, workspace, pipeline) |
| **TaskRun** | Individual task execution: script, code checksum, inputs, outputs, container, and dependencies |
| **FileOutput** | Output file: path, checksum, size, timestamp, and links back to the task and workflow that produced it |

> Each record gets a **lineage ID (LID)** — a `lid://` URI that uniquely identifies the entity. LIDs are navigable: every LID and lineage label is a clickable link that queries all related entities across your organization.

Reviewer 4: ```suggestion
Each record gets a lineage ID (LID), a `lid://` URI that uniquely identifies the entity. Every LID and lineage label renders as a clickable link, letting you navigate to all related entities across your organization.
```

```groovy
lineage.enabled = true
lineage.store.location = '<PATH_TO_STORAGE>'
```

Only runs executed with this setting generate lineage data. Runs without it display a note on the **Run Info** tab:

> *Lineage tracking was not enabled for this run. Add `lineage.enabled = true` to your Nextflow config to capture lineage data.*

## Data lineage displayed in Seqera Platform

### Workflow run details

When a run was executed with lineage enabled, the [run details page][run-details] displays lineage data across the following tabs:

**Run Info** — shows the lineage ID, lineage labels, and the full Platform context captured at execution time: user, workspace, compute environment, pipeline name, revision, and commit ID.

**Tasks** — displays the lineage ID and lineage labels for each `TaskRun` alongside existing task data, so you can trace any task back to its lineage record. All task file inputs and outputs, and upstream and downstream tasks linked by lineage records are displayed.

**Inputs** — lists all input datasets and parameters with file paths, types, and lineage IDs and lineage labels where available.

> **Outputs** — lists all `FileOutput` records linked to the workflow run: output name, file path, type, lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].

Reviewer 4: ```suggestion
- **Run Info**: Shows the lineage ID, lineage labels, and the full Platform context captured at execution time: user, workspace, compute environment, pipeline name, revision, and commit ID.
- **Tasks**: Displays the lineage ID and lineage labels for each `TaskRun` alongside existing task data, so you can trace any task back to its lineage record. All task file inputs and outputs, and upstream and downstream tasks linked by lineage records, are displayed.
- **Inputs**: Lists all input datasets and parameters with file paths, types, and lineage IDs and lineage labels where available.
- **Outputs**: Lists all `FileOutput` records linked to the workflow run: output name, file path, type, lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].
```

---
title: "Data Lineage"
description: "Using data lineage in Seqera Platform."
date created: "2026-05-04"
last updated: "2026-05-04"
tags: [data lineage data-lineage provenance governance reproducibility lineage-id lid label]
---

:::info
Data lineage in platform is currently in public preview.

Data lineage requires Nextflow v25.04 or later with either `lineage.enabled = true` in your Nextflow pipeline configuration (for per pipeline configuration) or defined in the **Lineage** workspace setting. The feature is experimental and subject to change.

> Please consult this guide for the latest information on recommended configuration and limitations.

Reviewer 4: ```suggestion
Data lineage in Platform is in public preview. It requires Nextflow v25.04 or later. Enable it per-pipeline (`lineage.enabled = true` in your Nextflow config) or workspace-wide via the **Lineage** workspace setting.

The feature is experimental and subject to change. See this guide for the latest configuration recommendations and limitations.
```

:::

> Data lineage tracks the full provenance of every pipeline run — what executed, what data was consumed, and what outputs were produced (both task-level and workflow-level). This allows auditing of pipeline results, verification of reproducibility, and tracing file provenance.

Reviewer 4: ```suggestion
Data lineage tracks the full provenance of every pipeline run at both the task and workflow level, including what executed, what data it consumed, and what outputs it produced. Use it to audit results, verify reproducibility, and trace file provenance.
```

## Why data lineage matters

> Production pipelines generate results that teams need to trust, audit, and reproduce. Data lineage answers the question "how exactly was this result produced?" with a precise, immutable record.

Reviewer 4: ```suggestion
Production pipelines generate results that teams need to trust, audit, and reproduce. Data lineage provides a precise, immutable record of how each result was produced.
```

- **Reproducibility**: Every run, task, and output file receives a unique **Lineage ID (LID)** — a traversable URI pointing to a structured record of exactly what ran. You can verify that two runs produced identical results, or identify precisely where they diverged.
- **Auditing and compliance**: For teams in regulated industries (pharma, clinical genomics, CROs), lineage provides the audit trail needed for regulatory compliance. Each record captures inputs, outputs, parameters, compute environment, and the user who launched the run.
- **Debugging**: When a cached task unexpectedly re-executes or a pipeline produces an unexpected result, lineage lets you trace backward from any output to all contributing tasks and parameters. You can compare two task runs to isolate exactly what changed.
- **Broader team access**: Previously, exploring Nextflow lineage required CLI access and comfort reading raw JSON. Lineage data is surfaced in both pipeline run details pages and Data Explorer, so users can inspect provenance directly.

> - **Cross-workflow discoverability**: [Workflow output labels][workflow-labels] make output files discoverable across runs. Rather than knowing which specific run produced a file, query lineage records by label to find all matching outputs workspace-wide.

Reviewer 4: ```suggestion
- **Reproducibility**: Every run, task, and output file receives a unique lineage ID (LID), a traversable URI that points to a structured record of what ran. Verify that two runs produced identical results, or identify where they diverged.
- **Auditing and compliance**: For teams in regulated industries such as pharma, clinical genomics, and CROs, lineage provides the audit trail needed for regulatory compliance. Each record captures inputs, outputs, parameters, compute environment, and the user who launched the run.
- **Debugging**: When a cached task unexpectedly re-executes, or a pipeline produces an unexpected result, lineage traces backward from any output to all contributing tasks and parameters. Compare two task runs to isolate what changed.
- **Broader team access**: Exploring Nextflow lineage previously required CLI access and comfort reading raw JSON. Platform now surfaces lineage data in pipeline run detail pages and Data Explorer. Users can inspect provenance directly.
- **Cross-workflow discoverability**: [Workflow output labels][workflow-labels] make output files discoverable across runs. Query lineage records by label to find all matching outputs workspace-wide, without knowing which specific run produced a file.
```

Reviewer 4: Need to confirm Lineage ID or lineage ID

Author: Pls use "lineage ID"

## How data lineage works

Nextflow creates a structured JSON record for each entity in your pipeline when lineage is enabled:

| Record type | Description |
|---|---|
| **WorkflowRun** | Full pipeline execution: repository, commit ID, parameters, compute environment, session ID, and Platform context (user, workspace, pipeline) |
| **TaskRun** | Individual task execution: script, code checksum, inputs, outputs, container, and dependencies |
| **FileOutput** | Output file: path, checksum, size, timestamp, and links back to the task and workflow that produced it |

> Each record gets a **Lineage ID (LID)** — a `lid://` URI that uniquely identifies the entity. LIDs are navigable: every LID and lineage label is a clickable link that queries all related entities across your organization.

Reviewer 4: ```suggestion
Each record gets a lineage ID (LID), a `lid://` URI that uniquely identifies the entity. Every LID and lineage label renders as a clickable link, letting you navigate to all related entities across your organization.
```

### Configure workspace settings

> Before collecting lineage data, configure the lineage storage location for your workspace. Go to **Workspace Settings** and open the [**Lineage** settings][workspace-lineage] to set the storage bucket and path where the lineage data is stored and indexed. This applies to **all** pipeline runs in the workspace.

Reviewer 4: ```suggestion
Before collecting lineage data, configure the lineage storage location for your workspace. Go to **Workspace Settings** and open the [**Lineage** settings][workspace-lineage] to set the storage bucket and path where lineage data is stored and indexed. This applies to all pipeline runs in the workspace.
```

:::danger

> Changing the lineage storage bucket path after runs have generated lineage data will result in historic data loss. The lineage index is tied to this location — changing it makes existing records inaccessible. If you need to move the lineage data storage location, first copy all existing lineage data to the new bucket and path (for example, `aws s3 cp --recursive s3://old-bucket/path s3://new-bucket/path`), then update the workspace setting.

Reviewer 4: ```suggestion
Changing the lineage storage bucket path after lineage data is generated results in historic data loss. The lineage index is tied to this location. Changing it makes existing records inaccessible. To move the storage location, first copy all existing lineage data to the new bucket and path (for example, `aws s3 cp --recursive s3://old-bucket/path s3://new-bucket/path`), then update the workspace setting.
```

:::

### Enable per pipeline lineage in Nextflow

> To test lineage within a single pipeline, add the following to your Nextflow configuration file before running your pipeline:

Reviewer 4: ```suggestion
To test lineage within a single pipeline, add the following to your Nextflow config file before running your pipeline:
```

```groovy
lineage.enabled = true
lineage.store.location = '<PATH_TO_STORAGE>'
```

> Only runs executed with this setting generate lineage data. Runs without it display a note on the Run Info tab:

Reviewer 4: ```suggestion
Only runs executed with this setting generate lineage data. Runs without it display a note on the **Run Info** tab:
```

> *Lineage tracking was not enabled for this run. Add `lineage.enabled = true` to your Nextflow config to capture lineage data.*

## Data lineage displayed in Seqera Platform

### Workflow run details

When a run was executed with lineage enabled, the [run details page][run-details] displays lineage data across the following tabs:

**Run Info** — shows the Lineage ID, lineage labels, and the full Platform context captured at execution time: user, workspace, compute environment, pipeline name, revision, and commit ID.

**Tasks** — displays the Lineage ID and lineage labels for each `TaskRun` alongside existing task data, so you can trace any task back to its lineage record. All task file inputs and outputs, and upstream and downstream tasks linked by lineage records are displayed.

**Inputs** — lists all input datasets and parameters with file paths, types, and Lineage IDs and lineage labels where available.

> **Outputs** — lists all `FileOutput` records linked to the workflow run: output name, file path, type, Lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].

Reviewer 2: ```suggestion
**Outputs** — lists all `FileOutput` records linked to the workflow run: output name, file path, type, lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].
```

Author: Fixed throughout

> **Outputs** — lists all `FileOutput` records linked to the workflow run: output name, file path, type, Lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].

Reviewer 2: "Lineage ID" should be lowercase for consistency's sake. I know we refer to "LID" and that can remain (acronyms don't need to be capitalized). @VictoriaAdamonis. tagging you in for any UI impact this may have.

Author: Fixed throughout

> **Outputs** — lists all `FileOutput` records linked to the workflow run: output name, file path, type, Lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].

Reviewer 4: ```suggestion
**Run Info**: Shows the Lineage ID, lineage labels, and the full Platform context captured at execution time: user, workspace, compute environment, pipeline name, revision, and commit ID.

**Tasks**: Displays the Lineage ID and lineage labels for each `TaskRun` alongside existing task data, so you can trace any task back to its lineage record. All task file inputs and outputs, and upstream and downstream tasks linked by lineage records, are displayed.

**Inputs**: Lists all input datasets and parameters with file paths, types, and Lineage IDs and lineage labels where available.

**Outputs**: Lists all `FileOutput` records linked to the workflow run: output name, file path, type, Lineage ID, and lineage labels. Files link directly to [Data Explorer][data-explorer].
```

:::tip

> All LIDs and lineage labels are clickable links. Clicking any LID opens the organization-level lineage search pre-filled with that identifier.

Reviewer 4: ```suggestion
All LIDs and lineage labels are clickable links. Click any LID to open the organization-level lineage search pre-filled with that identifier.
```

:::

### Data Explorer

> Output objects generated by a lineage-enabled run display the **LID** and any **lineage labels** when you preview the object in Data Explorer. This lets you trace any file back to the pipeline run that produced it.

Reviewer 4: ```suggestion
Output objects from a lineage-enabled run display their LID and any lineage labels when you preview the object in Data Explorer. You can trace any file back to the pipeline run that produced it.
```

## Lineage labels

Assign lineage labels to output files using the `label` directive in your Nextflow process definitions. Labels appear in lineage records and are searchable across your workspace.

> Both Seqera Platform labels and Nextflow lineage labels propagate to lineage records. Seqera Platform excludes **resource labels** — they relate to underlying compute resources, not the data itself.

Reviewer 4: ```suggestion
Assign lineage labels to output files using the `label` directive in your Nextflow process definitions. Labels appear in lineage records and are searchable across your workspace.

Both Seqera Platform labels and Nextflow lineage labels propagate to lineage records. Seqera Platform excludes resource labels as they relate to underlying compute resources, not the data itself.
```

:::info

> Nextflow lineage labels are **immutable** — they are set at execution time and cannot be changed. Seqera Platform labels are **mutable**. If you update Platform labels after a run completes, a mismatch between Platform run labels and lineage labels is possible. This is expected behavior.

Reviewer 4: ```suggestion
Nextflow lineage labels are immutable. They are set at execution time and cannot be changed. Seqera Platform labels are mutable. Updating Platform labels after a run completes can produce a mismatch between Platform run labels and lineage labels. This is expected behavior.
```
